File: engine/pong_engine/physics.py
from __future__ import annotations

from math import hypot

from pong_engine.config import (
    ARENA_HEIGHT,
    ARENA_WIDTH,
    BALL_ACCELERATION_FACTOR,
    BALL_INITIAL_SPEED,
    BALL_MAX_SPEED,
    BALL_SIZE,
    PADDLE_HEIGHT,
    PADDLE_SPEED,
    PADDLE_WIDTH,
)
from pong_engine.errors import InvalidActionError
from pong_engine.heuristics import ensure_rng
from pong_engine.state import Action, GameState, sample_launch_velocity


VALID_ACTIONS = {"up", "down", "none"}


def step(
    state: GameState,
    a_left: Action,
    a_right: Action,
    rng: object | None,
) -> GameState:
    ensure_rng(rng)
    validate_action(a_left)
    validate_action(a_right)

    paddle_left_y = move_paddle(state.paddle_left_y, a_left)
    paddle_right_y = move_paddle(state.paddle_right_y, a_right)

    next_x = state.ball_x + state.ball_vx
    next_y = state.ball_y + state.ball_vy
    next_vx = state.ball_vx
    next_vy = state.ball_vy
    next_speed = state.ball_speed

    if next_y < 0:
        next_y = -next_y
        next_vy = abs(next_vy)
    elif next_y + BALL_SIZE > ARENA_HEIGHT:
        overflow = next_y + BALL_SIZE - ARENA_HEIGHT
        next_y = ARENA_HEIGHT - BALL_SIZE - overflow
        next_vy = -abs(next_vy)

    if next_vx < 0 and collides_with_left_paddle(next_x, next_y, paddle_left_y):
        next_x = PADDLE_WIDTH
        next_speed = min(state.ball_speed * BALL_ACCELERATION_FACTOR, BALL_MAX_SPEED)
        next_vx, next_vy = rescale_velocity(abs(next_vx), next_vy, next_speed)
    elif next_vx > 0 and collides_with_right_paddle(next_x, next_y, paddle_right_y):
        next_x = ARENA_WIDTH - PADDLE_WIDTH - BALL_SIZE
        next_speed = min(state.ball_speed * BALL_ACCELERATION_FACTOR, BALL_MAX_SPEED)
        reflected_vx, next_vy = rescale_velocity(abs(next_vx), next_vy, next_speed)
        next_vx = -reflected_vx

    if next_x < 0:
        return create_reset_state(state, paddle_left_y, paddle_right_y, "right", rng)
    if next_x + BALL_SIZE > ARENA_WIDTH:
        return create_reset_state(state, paddle_left_y, paddle_right_y, "left", rng)

    return GameState(
        ball_x=next_x,
        ball_y=next_y,
        ball_vx=next_vx,
        ball_vy=next_vy,
        ball_speed=next_speed,
        paddle_left_y=paddle_left_y,
        paddle_right_y=paddle_right_y,
        score_left=state.score_left,
        score_right=state.score_right,
        tick=state.tick + 1,
    )


def validate_action(action: Action) -> None:
    if action not in VALID_ACTIONS:
        raise InvalidActionError(f"Invalid action: {action}")


def move_paddle(current_y: int, action: Action) -> int:
    delta = 0
    if action == "up":
        delta = -PADDLE_SPEED
    elif action == "down":
        delta = PADDLE_SPEED
    return clamp(current_y + delta, 0, ARENA_HEIGHT - PADDLE_HEIGHT)


def collides_with_left_paddle(ball_x: float, ball_y: float, paddle_y: int) -> bool:
    return ball_x <= PADDLE_WIDTH and overlaps_paddle(ball_y, paddle_y)


def collides_with_right_paddle(ball_x: float, ball_y: float, paddle_y: int) -> bool:
    paddle_x = ARENA_WIDTH - PADDLE_WIDTH
    return ball_x + BALL_SIZE >= paddle_x and overlaps_paddle(ball_y, paddle_y)


def overlaps_paddle(ball_y: float, paddle_y: int) -> bool:
    return ball_y + BALL_SIZE > paddle_y and ball_y < paddle_y + PADDLE_HEIGHT


def rescale_velocity(vx: float, vy: float, speed: float) -> tuple[float, float]:
    magnitude = hypot(vx, vy)
    if magnitude == 0:
        return speed, 0.0
    scale = speed / magnitude
    return vx * scale, vy * scale


def create_reset_state(
    state: GameState,
    paddle_left_y: int,
    paddle_right_y: int,
    scorer: str,
    rng: object | None,
) -> GameState:
    ball_vx, ball_vy = sample_launch_velocity(BALL_INITIAL_SPEED, rng)
    score_left = state.score_left + (1 if scorer == "left" else 0)
    score_right = state.score_right + (1 if scorer == "right" else 0)
    return GameState(
        ball_x=(ARENA_WIDTH - BALL_SIZE) / 2,
        ball_y=(ARENA_HEIGHT - BALL_SIZE) / 2,
        ball_vx=ball_vx,
        ball_vy=ball_vy,
        ball_speed=BALL_INITIAL_SPEED,
        paddle_left_y=paddle_left_y,
        paddle_right_y=paddle_right_y,
        score_left=score_left,
        score_right=score_right,
        tick=state.tick + 1,
    )


def clamp(value: int, minimum: int, maximum: int) -> int:
    return max(minimum, min(value, maximum))

File: web/src/engine/physics.ts
import {
  ARENA_HEIGHT,
  ARENA_WIDTH,
  BALL_ACCELERATION_FACTOR,
  BALL_INITIAL_SPEED,
  BALL_MAX_SPEED,
  BALL_SIZE,
  PADDLE_HEIGHT,
  PADDLE_SPEED,
  PADDLE_WIDTH,
} from "@/engine/config";
import { InvalidActionError } from "@/engine/errors";
import { assertRng, type Rng } from "@/engine/rng";
import { sampleLaunchVelocity, type Action, type GameState } from "@/engine/state";

const VALID_ACTIONS = new Set<Action>(["up", "down", "none"]);

export function step(
  state: GameState,
  aLeft: Action,
  aRight: Action,
  rng: Rng | null | undefined,
): GameState {
  assertRng(rng);
  validateAction(aLeft);
  validateAction(aRight);

  const paddleLeftY = movePaddle(state.paddleLeftY, aLeft);
  const paddleRightY = movePaddle(state.paddleRightY, aRight);

  let nextX = state.ballX + state.ballVx;
  let nextY = state.ballY + state.ballVy;
  let nextVx = state.ballVx;
  let nextVy = state.ballVy;
  let nextSpeed = state.ballSpeed;

  if (nextY < 0) {
    nextY = -nextY;
    nextVy = Math.abs(nextVy);
  } else if (nextY + BALL_SIZE > ARENA_HEIGHT) {
    const overflow = nextY + BALL_SIZE - ARENA_HEIGHT;
    nextY = ARENA_HEIGHT - BALL_SIZE - overflow;
    nextVy = -Math.abs(nextVy);
  }

  if (nextVx < 0 && collidesWithLeftPaddle(nextX, nextY, paddleLeftY)) {
    nextX = PADDLE_WIDTH;
    nextSpeed = Math.min(state.ballSpeed * BALL_ACCELERATION_FACTOR, BALL_MAX_SPEED);
    [nextVx, nextVy] = rescaleVelocity(Math.abs(nextVx), nextVy, nextSpeed);
  } else if (nextVx > 0 && collidesWithRightPaddle(nextX, nextY, paddleRightY)) {
    nextX = ARENA_WIDTH - PADDLE_WIDTH - BALL_SIZE;
    nextSpeed = Math.min(state.ballSpeed * BALL_ACCELERATION_FACTOR, BALL_MAX_SPEED);
    const [reflectedVx, reflectedVy] = rescaleVelocity(Math.abs(nextVx), nextVy, nextSpeed);
    nextVx = -reflectedVx;
    nextVy = reflectedVy;
  }

  if (nextX < 0) {
    return createResetState(state, paddleLeftY, paddleRightY, "right", rng);
  }
  if (nextX + BALL_SIZE > ARENA_WIDTH) {
    return createResetState(state, paddleLeftY, paddleRightY, "left", rng);
  }

  return {
    ballX: nextX,
    ballY: nextY,
    ballVx: nextVx,
    ballVy: nextVy,
    ballSpeed: nextSpeed,
    paddleLeftY,
    paddleRightY,
    scoreLeft: state.scoreLeft,
    scoreRight: state.scoreRight,
    tick: state.tick + 1,
  };
}

function validateAction(action: Action): void {
  if (!VALID_ACTIONS.has(action)) {
    throw new InvalidActionError(`Invalid action: ${action}`);
  }
}

function movePaddle(currentY: number, action: Action): number {
  let delta = 0;
  if (action === "up") {
    delta = -PADDLE_SPEED;
  } else if (action === "down") {
    delta = PADDLE_SPEED;
  }
  return clamp(currentY + delta, 0, ARENA_HEIGHT - PADDLE_HEIGHT);
}

function collidesWithLeftPaddle(ballX: number, ballY: number, paddleY: number): boolean {
  return ballX <= PADDLE_WIDTH && overlapsPaddle(ballY, paddleY);
}

function collidesWithRightPaddle(ballX: number, ballY: number, paddleY: number): boolean {
  const paddleX = ARENA_WIDTH - PADDLE_WIDTH;
  return ballX + BALL_SIZE >= paddleX && overlapsPaddle(ballY, paddleY);
}

function overlapsPaddle(ballY: number, paddleY: number): boolean {
  return ballY + BALL_SIZE > paddleY && ballY < paddleY + PADDLE_HEIGHT;
}

function rescaleVelocity(vx: number, vy: number, speed: number): [number, number] {
  const magnitude = Math.hypot(vx, vy);
  if (magnitude === 0) {
    return [speed, 0];
  }
  const scale = speed / magnitude;
  return [vx * scale, vy * scale];
}

function createResetState(
  state: GameState,
  paddleLeftY: number,
  paddleRightY: number,
  scorer: "left" | "right",
  rng: Rng,
): GameState {
  const [ballVx, ballVy] = sampleLaunchVelocity(BALL_INITIAL_SPEED, rng);
  return {
    ballX: (ARENA_WIDTH - BALL_SIZE) / 2,
    ballY: (ARENA_HEIGHT - BALL_SIZE) / 2,
    ballVx,
    ballVy,
    ballSpeed: BALL_INITIAL_SPEED,
    paddleLeftY,
    paddleRightY,
    scoreLeft: state.scoreLeft + (scorer === "left" ? 1 : 0),
    scoreRight: state.scoreRight + (scorer === "right" ? 1 : 0),
    tick: state.tick + 1,
  };
}

function clamp(value: number, minimum: number, maximum: number): number {
  return Math.max(minimum, Math.min(value, maximum));
}
